### ISSUE TITLE
Plotting 3D events data with an extent of 2 fails

### ISSUE THREAD

**Issue by SimonHeybrock** [scipp/scipp#1557]
```python
import scipp as sc
import numpy as np
data = sc.neutron.load(filename='some_event_data.nxs')
data = sc.concatenate(data, data+data, 'run')
data.plot()
```
Now drag the slider to get the following exception:

```python
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
~/software/miniconda3/envs/scipp-developer/lib/python3.7/site-packages/ipywidgets/widgets/widget.py in _handle_msg(self, msg)
    674                 if 'buffer_paths' in data:
    675                     _put_buffers(state, data['buffer_paths'], msg['buffers'])
--> 676                 self.set_state(state)
    677 
    678         # Handle a state request.

~/software/miniconda3/envs/scipp-developer/lib/python3.7/site-packages/ipywidgets/widgets/widget.py in set_state(self, sync_data)
    543                     from_json = self.trait_metadata(name, 'from_json',
    544                                                     self._trait_from_json)
--> 545                     self.set_trait(name, from_json(sync_data[name], self))
    546 
    547     def send(self, content, buffers=None):

~/software/miniconda3/envs/scipp-developer/lib/python3.7/contextlib.py in __exit__(self, type, value, traceback)
    117         if type is None:
    118             try:
--> 119                 next(self.gen)
    120             except StopIteration:
    121                 return False

~/software/miniconda3/envs/scipp-developer/lib/python3.7/site-packages/traitlets/traitlets.py in hold_trait_notifications(self)
   1129                 for changes in cache.values():
   1130                     for change in changes:
-> 1131                         self.notify_change(change)
   1132 
   1133     def _notify_trait(self, name, old_value, new_value):

~/software/miniconda3/envs/scipp-developer/lib/python3.7/site-packages/ipywidgets/widgets/widget.py in notify_change(self, change)
    604                 # Send new state to front-end
    605                 self.send_state(key=name)
--> 606         super(Widget, self).notify_change(change)
    607 
    608     def __repr__(self):

~/software/miniconda3/envs/scipp-developer/lib/python3.7/site-packages/traitlets/traitlets.py in notify_change(self, change)
   1174                 c = getattr(self, c.name)
   1175 
-> 1176             c(change)
   1177 
   1178     def _add_notifiers(self, handler, name, type):

~/software/scipp/scipp/plot/controller.py in update_data(self, change)
    435             owner_dim = self.widgets.get_index_dim(change["owner"].index)
    436             lower, upper = self.model.get_slice_coord_bounds(
--> 437                 self.name, owner_dim, slices[owner_dim])
    438             self.widgets.update_slider_readout(change["owner"].index, lower,
    439                                                upper, slices[owner_dim],

~/software/scipp/scipp/plot/model.py in get_slice_coord_bounds(self, name, dim, bounds)
    240             dim,
    241             bounds[0]].value, self.data_arrays[name].meta[dim][dim,
--> 242                                                                bounds[1]].value
    243 
    244     def get_data_names(self):

RuntimeError: The requested index 2 is out of range. Dimension size is 2 and the allowed range is [-2:1].
```

Does not seem to happend with dense data

**Comment by SimonHeybrock** [comment]
I think the problem may be a hack in `resampling_model.py`, where a temporary edge coord is changed back to a non-edge coord.

**Comment by Matthew-Andrew** [comment]
Now it is causing issue do we want to resolve this by solving #1469 ? Or do we want to improve the temporary hack so it handles this edge case?

**Comment by Matthew-Andrew** [comment]
A related error is that if you run the script above and then instead of moving the run slider switch to `spectrum` being the z dimension you get the error message below. I think this is because  run does not exist as a coord in the original DataArray so is created to be added to the PlotArray object. The ResamplingModel still references the original array however.

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
~/scipp/install/scipp/plot/widgets.py in update_buttons(self, owner)
    223         self.unit_labels[new_ind].value = self.interface["get_coord_unit"](
    224             new_dim)
--> 225         self.interface["swap_dimensions"](new_ind, old_dim, new_dim)
    226 
    227     def get_index_dim(self, index):

~/scipp/install/scipp/plot/controller.py in swap_dimensions(self, index, old_dim, new_dim)
    363         self.axes[key] = old_dim
    364         self.axes[index] = new_dim
--> 365         self.update_axes()
    366         self.update_log_axes_buttons()
    367         # Update the slider readout here because the widgets do not have access

~/scipp/install/scipp/plot/controller.py in update_axes(self, change, normalize)
    409         if self.profile is not None:
    410             self.toggle_profile_view()
--> 411         self.update_data()
    412         if normalize:
    413             self.rescale_to_data()

~/scipp/install/scipp/plot/controller.py in update_data(self, change)
    439 
    440         new_values = self.model.update_data(slices,
--> 441                                             mask_info=self.get_masks_info())
    442         self.view.update_data(new_values, info=info)
    443         if self.panel is not None:

~/scipp/install/scipp/plot/model2d.py in update_data(self, slices, mask_info)
     95                 self._model.resolution[dim] = 1
     96                 self._model.bounds[dim] = (start, stop)
---> 97         return self._update_image(mask_info=mask_info)
     98 
     99     def update_viewport(self, xylims, mask_info):

~/scipp/install/scipp/plot/model2d.py in _update_image(self, extent, mask_info)
     75         Resample 2d images to a fixed resolution to handle very large images.
     76         """
---> 77         data = self._model.data
     78         for dim in self._squeeze:
     79             data = data[dim, 0]

~/scipp/install/scipp/plot/resampling_model.py in data(self)
     37     @property
     38     def data(self):
---> 39         self._call_resample()
     40         return self._resampled
     41 

~/scipp/install/scipp/plot/resampling_model.py in _call_resample(self)
    115             self._resampled_params = params
    116             self._edges = self._make_edges(params)
--> 117             self._resampled = self._resample(out)
    118         if self._home is None:
    119             self._home = self._resampled

~/scipp/install/scipp/plot/resampling_model.py in _resample(self, array)
    149         binned = sc.bin_with_coords(array.data, array.meta,
    150                                     self.edges[:-1] + [bounds], [])
--> 151         a = sc.histogram(binned, edges)
    152         for name, mask in array.masks.items():
    153             a.masks[name] = self._rebin(mask, array.meta)

RuntimeError: Expected <scipp.ConstView>
  pulse-time:  <scipp.VariableView>      int64      [ns]             (event)  []
  tof:  <scipp.VariableView>      float64    [μs]            (event)  []
 to contain run.
```

**Comment by Matthew-Andrew** [comment]
A related error is that if you run the script above and then instead of moving the run slider switch to `spectrum` being the z dimension you get the error message below. I think this is because  run does not exist as a coord in the original DataArray so is created to be added to the PlotArray object. The ResamplingModel still references the original array however.

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
~/scipp/install/scipp/plot/widgets.py in update_buttons(self, owner)
    223         self.unit_labels[new_ind].value = self.interface["get_coord_unit"](
    224             new_dim)
--> 225         self.interface["swap_dimensions"](new_ind, old_dim, new_dim)
    226 
    227     def get_index_dim(self, index):

~/scipp/install/scipp/plot/controller.py in swap_dimensions(self, index, old_dim, new_dim)
    363         self.axes[key] = old_dim
    364         self.axes[index] = new_dim
--> 365         self.update_axes()
    366         self.update_log_axes_buttons()
    367         # Update the slider readout here because the widgets do not have access

~/scipp/install/scipp/plot/controller.py in update_axes(self, change, normalize)
    409         if self.profile is not None:
    410             self.toggle_profile_view()
--> 411         self.update_data()
    412         if normalize:
    413             self.rescale_to_data()

~/scipp/install/scipp/plot/controller.py in update_data(self, change)
    439 
    440         new_values = self.model.update_data(slices,
--> 441                                             mask_info=self.get_masks_info())
    442         self.view.update_data(new_values, info=info)
    443         if self.panel is not None:

~/scipp/install/scipp/plot/model2d.py in update_data(self, slices, mask_info)
     95                 self._model.resolution[dim] = 1
     96                 self._model.bounds[dim] = (start, stop)
---> 97         return self._update_image(mask_info=mask_info)
     98 
     99     def update_viewport(self, xylims, mask_info):

~/scipp/install/scipp/plot/model2d.py in _update_image(self, extent, mask_info)
     75         Resample 2d images to a fixed resolution to handle very large images.
     76         """
---> 77         data = self._model.data
     78         for dim in self._squeeze:
     79             data = data[dim, 0]

~/scipp/install/scipp/plot/resampling_model.py in data(self)
     37     @property
     38     def data(self):
---> 39         self._call_resample()
     40         return self._resampled
     41 

~/scipp/install/scipp/plot/resampling_model.py in _call_resample(self)
    115             self._resampled_params = params
    116             self._edges = self._make_edges(params)
--> 117             self._resampled = self._resample(out)
    118         if self._home is None:
    119             self._home = self._resampled

~/scipp/install/scipp/plot/resampling_model.py in _resample(self, array)
    149         binned = sc.bin_with_coords(array.data, array.meta,
    150                                     self.edges[:-1] + [bounds], [])
--> 151         a = sc.histogram(binned, edges)
    152         for name, mask in array.masks.items():
    153             a.masks[name] = self._rebin(mask, array.meta)

RuntimeError: Expected <scipp.ConstView>
  pulse-time:  <scipp.VariableView>      int64      [ns]             (event)  []
  tof:  <scipp.VariableView>      float64    [μs]            (event)  []
 to contain run.
```

**Comment by SimonHeybrock** [comment]
> Now it is causing issue do we want to resolve this by solving #1469 ? Or do we want to improve the temporary hack so it handles this edge case?

Depends on what is easier.

> A related error is that if you run the script above and then instead of moving the run slider switch to `spectrum` being the z dimension you get the error message below.
> 
> ```
> ~/scipp/install/scipp/plot/resampling_model.py in _resample(self, array)
>     149         binned = sc.bin_with_coords(array.data, array.meta,
>     150                                     self.edges[:-1] + [bounds], [])
> --> 151         a = sc.histogram(binned, edges)
>     152         for name, mask in array.masks.items():
>     153             a.masks[name] = self._rebin(mask, array.meta)
> 
> RuntimeError: Expected <scipp.ConstView>
>   pulse-time:  <scipp.VariableView>      int64      [ns]             (event)  []
>   tof:  <scipp.VariableView>      float64    [μs]            (event)  []
>  to contain run.
> ```

I see why this is happening (missing event coord), the solution would be to not use `histogram` for the inner edges in this case. Do you think I could/should take over this issue, it may be easier? Or we can go through it together if you like.

**Comment by Matthew-Andrew** [comment]
Happy for you to take it over if you want. From what I can currently tell the original issue is due to the fact that the ResamplingBinnedModel requires non-edge coords but the PlotModel needs edge coords and they share and modify the same PlotArray instance. I would expect this interaction to cause multiple problems along the lines of the one described here depending on ordering. I'm currently trying to understand the context of the surrounding code enough to work out what change won't break other things so someone already more familiar with the plotting could definitely do it faster.

**Comment by SimonHeybrock** [comment]
Fixed in #1574.
